File: crates/vici/src/pending.rs
//! The pending-input parser: vi's command grammar.
//!
//! A trie alone cannot express vi's normal-mode syntax, because counts and
//! operators *compose* rather than sequence:
//!
//! ```text
//! [count] operator [count] motion | textobject
//! [count] command
//! ```
//!
//! So this type holds the compositional state — counts, a pending operator, a
//! pending object scope, a pending character argument — and defers the rest to
//! [`Keymap`]. It touches no buffer, which makes the whole grammar testable as
//! string in, [`Command`] out.

use crate::command::{AwaitChar, Command, Mode, Motion, ObjectScope, Operator, Target, TextObject};
use crate::key::{Key, KeyCode};
use crate::keymap::{Binding, Keymap, Layer, Walk};
use crate::motion::Find;

/// What feeding one key produced.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Resolution {
    /// A valid partial sequence. Show the keys as a `showcmd` hint if you like.
    Pending,
    /// A complete command.
    Command {
        command: Command,
        /// The effective count, `None` when the user gave none.
        ///
        /// Distinguishing `None` from `Some(1)` matters: `G` with no count goes to
        /// the last row, `1G` to the first.
        count: Option<usize>,
        /// Every key consumed, in order. This is what dot-repeat and macro
        /// recording store — replaying keys rather than re-executing commands is
        /// what makes both fall out of one mechanism.
        keys: Vec<Key>,
    },
    /// Not a valid sequence. State has been reset; the host may beep.
    Rejected { keys: Vec<Key> },
    /// `<Esc>` abandoned a partial sequence.
    Cancelled { keys: Vec<Key> },
}

/// Accumulated state between keys.
#[derive(Debug, Clone, Default)]
pub struct Pending {
    keys: Vec<Key>,
    /// Start of the partial path through the current layer's bindings.
    path_start: usize,
    count_before: Option<usize>,
    count_after: Option<usize>,
    operator: Option<Operator>,
    scope: Option<ObjectScope>,
    awaiting: Option<AwaitChar>,
    search: Option<SearchInput>,
}

#[derive(Debug, Clone)]
struct SearchInput {
    pattern: String,
    backward: bool,
}

impl Pending {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Nothing accumulated — the next key starts a fresh command.
    #[must_use]
    pub fn is_idle(&self) -> bool {
        self.path_start == self.keys.len()
            && self.count_before.is_none()
            && self.count_after.is_none()
            && self.operator.is_none()
            && self.scope.is_none()
            && self.awaiting.is_none()
            && self.search.is_none()
    }

    /// Keys consumed so far, for a `showcmd` indicator.
    #[must_use]
    pub fn keys(&self) -> &[Key] {
        &self.keys
    }

    /// The operator awaiting a target, if any.
    #[must_use]
    pub fn operator(&self) -> Option<Operator> {
        self.operator
    }

    pub fn reset(&mut self) {
        *self = Self::new();
    }

    /// Feed one key.
    ///
    /// `mode` selects the grammar: [`Mode::is_command`] modes get counts and
    /// operators, insert and replace get a straight keymap lookup with unbound
    /// printable keys falling through to text.
    pub fn feed(&mut self, key: Key, mode: Mode, keymap: &Keymap) -> Resolution {
        if !mode.is_command() {
            self.keys.push(key);
            return self.feed_insert(key, keymap);
        }

        // `<Esc>` abandons a partial sequence. When nothing is pending it falls
        // through to the keymap, where it means "return to normal mode".
        let idle = self.is_idle();
        let path_is_empty = self.path_start == self.keys.len();
        self.keys.push(key);
        if key.code == KeyCode::Esc && !idle {
            return self.cancel();
        }

        if self.search.is_some() {
            return self.feed_search(key);
        }

        if let Some(awaiting) = self.awaiting {
            return match key.as_text() {
                Some(ch) => self.resolve_await(awaiting, ch),
                None => self.reject(),
            };
        }

        if let Some(scope) = self.scope {
            return match keymap.object(key) {
                Some(object) => self.resolve_object(scope, object),
                None => self.reject(),
            };
        }

        // Counts accumulate only at the start of a key path, and a leading `0` is
        // the first-column motion rather than a digit — the classic ambiguity.
        if path_is_empty && let Some(digit) = key.as_digit() {
            let digit = digit as usize;
            let slot = if self.operator.is_some() {
                &mut self.count_after
            } else {
                &mut self.count_before
            };
            if !(digit == 0 && slot.is_none()) {
                *slot = Some(slot.unwrap_or(0).saturating_mul(10).saturating_add(digit));
                self.path_start = self.keys.len();
                return Resolution::Pending;
            }
        }

        let layer = if self.operator.is_some() {
            Layer::Operator
        } else {
            Layer::of(mode)
        };
        match keymap.walk(layer, &self.keys[self.path_start..]) {
            Walk::Prefix => Resolution::Pending,
            Walk::Unbound => self.reject(),
            Walk::Bound(binding) => self.apply(binding, mode),
        }
    }

    fn feed_insert(&mut self, key: Key, keymap: &Keymap) -> Resolution {
        match keymap.walk(Layer::Insert, &self.keys[self.path_start..]) {
            Walk::Prefix => Resolution::Pending,
            Walk::Bound(Binding::Command(command)) => self.finish(command),
            // Operators and object scopes are meaningless while inserting.
            Walk::Bound(_) => self.reject(),
            Walk::Unbound => {
                // An unbound printable key is text.
                if self.keys.len() - self.path_start == 1
                    && let Some(ch) = key.as_text()
                {
                    self.finish(Command::InsertText(ch))
                } else {
                    self.reject()
                }
            }
        }
    }

    fn apply(&mut self, binding: Binding, mode: Mode) -> Resolution {
        self.path_start = self.keys.len();
        match binding {
            Binding::Command(command) => {
                // An operator needs a target; a plain command is not one.
                if self.operator.is_some() {
                    return self.reject();
                }
                self.finish(command)
            }
            Binding::Motion(motion) => self.finish(self.with_motion(motion)),
            Binding::Operator(operator) => {
                // In visual mode an operator applies to the selection at once.
                if mode.is_visual() {
                    return self.finish(Command::Operate {
                        operator,
                        target: Target::Selection,
                    });
                }
                match self.operator {
                    None => {
                        self.operator = Some(operator);
                        Resolution::Pending
                    }
                    // A doubled operator is linewise: `dd`, `cc`, `yy`.
                    Some(active) if active == operator => self.finish(Command::Operate {
                        operator,
                        target: Target::CurrentRow,
                    }),
                    Some(_) => self.reject(),
                }
            }
            Binding::ObjectScope(scope) => {
                if self.operator.is_none() && !mode.is_visual() {
                    return self.reject();
                }
                self.scope = Some(scope);
                Resolution::Pending
            }
            Binding::Await(awaiting) => {
                if awaiting == AwaitChar::SurroundTarget
                    && !matches!(self.operator, Some(Operator::Change | Operator::Delete))
                {
                    // `Operator::Yank` is where `ys` can hook in when its target
                    // grammar is built. It is deliberately not part of surround's
                    // character-await path.
                    return self.reject();
                }
                self.awaiting = Some(awaiting);
                Resolution::Pending
            }
            Binding::Search { backward } => {
                self.search = Some(SearchInput {
                    pattern: String::new(),
                    backward,
                });
                Resolution::Pending
            }
        }
    }

    fn feed_search(&mut self, key: Key) -> Resolution {
        match key.code {
            KeyCode::Enter if key.mods.is_empty() => {
                let Some(search) = self.search.take() else {
                    return self.reject();
                };
                if search.pattern.is_empty() {
                    self.reject()
                } else {
                    self.finish(self.with_motion(Motion::Search {
                        pattern: search.pattern,
                        backward: search.backward,
                    }))
                }
            }
            KeyCode::Backspace if key.mods.is_empty() => {
                let Some(search) = self.search.as_mut() else {
                    return self.reject();
                };
                if search.pattern.pop().is_none() {
                    self.cancel()
                } else {
                    // `keys` doubles as showcmd. Keep its displayed pattern in
                    // step with editing; replaying this normalized spelling has
                    // the same result as replaying the literal backspace.
                    self.keys.pop();
                    self.keys.pop();
                    Resolution::Pending
                }
            }
            _ => match key.as_text() {
                Some(ch) => {
                    let Some(search) = self.search.as_mut() else {
                        return self.reject();
                    };
                    search.pattern.push(ch);
                    Resolution::Pending
                }
                None => self.reject(),
            },
        }
    }

    fn resolve_await(&mut self, awaiting: AwaitChar, ch: char) -> Resolution {
        self.awaiting = None;
        match awaiting {
            AwaitChar::Find { backward, till } => {
                let motion = Motion::Find(Find {
                    target: ch,
                    backward,
                    till,
                });
                self.finish(self.with_motion(motion))
            }
            // These take no target, so a pending operator is a syntax error.
            AwaitChar::ReplaceChar if self.operator.is_none() => {
                self.finish(Command::ReplaceChar(ch))
            }
            AwaitChar::RecordMacro if self.operator.is_none() => {
                self.finish(Command::RecordMacro(ch))
            }
            AwaitChar::PlayMacro if self.operator.is_none() => self.finish(Command::PlayMacro(ch)),
            AwaitChar::SetMark if self.operator.is_none() && ch.is_ascii_lowercase() => {
                self.finish(Command::SetMark(ch))
            }
            // `'` and `` ` `` name the position before the latest jump. They go
            // through the motion path like any other mark, so `d''` is linewise
            // and ``` d`` ``` characterwise, as `d'a` and `` d`a `` are.
            AwaitChar::GotoMark { exact }
                if ch.is_ascii_lowercase()
                    || matches!(ch, '<' | '>' | '[' | ']' | '^' | '\'' | '`') =>
            {
                self.finish(self.with_motion(Motion::Mark { name: ch, exact }))
            }
            AwaitChar::SurroundTarget => match self.operator {
                Some(Operator::Delete) => self.finish(Command::DeleteSurround(ch)),
                Some(Operator::Change) => {
                    self.awaiting = Some(AwaitChar::SurroundTo(ch));
                    Resolution::Pending
                }
                _ => self.reject(),
            },
            AwaitChar::SurroundTo(from) if self.operator == Some(Operator::Change) => {
                self.finish(Command::ChangeSurround { from, to: ch })
            }
            AwaitChar::SurroundSelection if self.operator.is_none() => {
                self.finish(Command::SurroundSelection(ch))
            }
            _ => self.reject(),
        }
    }

    fn resolve_object(&mut self, scope: ObjectScope, object: TextObject) -> Resolution {
        self.scope = None;
        let command = match self.operator {
            Some(operator) => Command::Operate {
                operator,
                target: Target::Object { scope, object },
            },
            None => Command::SelectObject { scope, object },
        };
        self.finish(command)
    }

    /// A motion is a movement on its own, or an operator's target.
    fn with_motion(&self, motion: Motion) -> Command {
        match self.operator {
            Some(operator) => Command::Operate {
                operator,
                target: Target::Motion(motion),
            },
            None => Command::Move(motion),
        }
    }

    /// vi multiplies the two counts: `2d3w` deletes six words.
    fn count(&self) -> Option<usize> {
        match (self.count_before, self.count_after) {
            (None, None) => None,
            (before, after) => Some(before.unwrap_or(1).saturating_mul(after.unwrap_or(1))),
        }
    }

    fn finish(&mut self, command: Command) -> Resolution {
        let count = self.count();
        let keys = core::mem::take(&mut self.keys);
        self.reset();
        Resolution::Command {
            command,
            count,
            keys,
        }
    }

    fn reject(&mut self) -> Resolution {
        let keys = core::mem::take(&mut self.keys);
        self.reset();
        Resolution::Rejected { keys }
    }

    fn cancel(&mut self) -> Resolution {
        let keys = core::mem::take(&mut self.keys);
        self.reset();
        Resolution::Cancelled { keys }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::command::{InsertAt, VisualKind};
    use crate::key::keys;

    fn resolve_in(mode: Mode, spec: &str) -> Resolution {
        let keymap = Keymap::vim();
        let mut pending = Pending::new();
        let parsed = keys(spec).expect("valid test notation");
        let (last, rest) = parsed.split_last().expect("non-empty test sequence");
        for key in rest {
            assert_eq!(pending.feed(*key, mode, &keymap), Resolution::Pending);
        }
        pending.feed(*last, mode, &keymap)
    }

    fn command(spec: &str) -> (Command, Option<usize>) {
        match resolve_in(Mode::Normal, spec) {
            Resolution::Command { command, count, .. } => (command, count),
            other => panic!("{spec:?} should resolve, got {other:?}"),
        }
    }

    #[test]
    fn counts_and_operator_composition() {
        let word = Target::Motion(Motion::WordForward { big: false });
        for (spec, expected, count) in [
            ("3j", Command::Move(Motion::Down), Some(3)),
            (
                "2d3w",
                Command::Operate {
                    operator: Operator::Delete,
                    target: word,
                },
                Some(6),
            ),
            (
                "diw",
                Command::Operate {
                    operator: Operator::Delete,
                    target: Target::Object {
                        scope: ObjectScope::Inner,
                        object: TextObject::Word { big: false },
                    },
                },
                None,
            ),
            (
                "dt;",
                Command::Operate {
                    operator: Operator::Delete,
                    target: Target::Motion(Motion::Find(Find {
                        target: ';',
                        backward: false,
                        till: true,
                    })),
                },
                None,
            ),
            (
                "d'a",
                Command::Operate {
                    operator: Operator::Delete,
                    target: Target::Motion(Motion::Mark {
                        name: 'a',
                        exact: false,
                    }),
                },
                None,
            ),
        ] {
            assert_eq!(command(spec), (expected, count), "{spec}");
        }
    }

    #[test]
    fn doubled_operators_are_linewise() {
        for (spec, operator) in [
            ("dd", Operator::Delete),
            ("cc", Operator::Change),
            ("yy", Operator::Yank),
            (">>", Operator::ShiftRight),
            ("<lt><lt>", Operator::ShiftLeft),
        ] {
            assert_eq!(
                command(spec),
                (
                    Command::Operate {
                        operator,
                        target: Target::CurrentRow,
                    },
                    None,
                )
            );
        }
    }

    #[test]
    fn surround_chains_character_awaits_without_dropping_the_operator() {
        let keymap = Keymap::vim();
        let mut pending = Pending::new();
        for spec in ["c", "s", "("] {
            let key = keys(spec).expect("valid test notation")[0];
            assert_eq!(
                pending.feed(key, Mode::Normal, &keymap),
                Resolution::Pending,
                "{spec}"
            );
            assert_eq!(pending.operator(), Some(Operator::Change), "{spec}");
        }
        assert_eq!(
            pending.feed(Key::char('{'), Mode::Normal, &keymap),
            Resolution::Command {
                command: Command::ChangeSurround { from: '(', to: '{' },
                count: None,
                keys: keys("cs({").expect("valid test notation"),
            }
        );

        assert_eq!(command("ds(").0, Command::DeleteSurround('('),);
        assert!(matches!(
            resolve_in(Mode::Normal, "ys"),
            Resolution::Rejected { .. }
        ));
        assert_eq!(
            resolve_in(Mode::Visual(VisualKind::Char), "S)"),
            Resolution::Command {
                command: Command::SurroundSelection(')'),
                count: None,
                keys: keys("S)").expect("valid test notation"),
            }
        );
    }

    #[test]
    fn cancellation_and_rejection_reset_the_parser() {
        for spec in ["d!", "dy", "dx", "diz", "f<Left>", "mA", "`A"] {
            assert!(matches!(
                resolve_in(Mode::Normal, spec),
                Resolution::Rejected { .. }
            ));
        }
        for spec in ["d<Esc>", "2d<Esc>", "f<Esc>", "/half<Esc>"] {
            assert!(matches!(
                resolve_in(Mode::Normal, spec),
                Resolution::Cancelled { .. }
            ));
        }
    }

    #[test]
    fn search_awaits_an_editable_line() {
        assert_eq!(
            command("3/foo<CR>"),
            (
                Command::Move(Motion::Search {
                    pattern: "foo".to_owned(),
                    backward: false,
                }),
                Some(3),
            )
        );
        assert_eq!(
            command("d?bar<CR>").0,
            Command::Operate {
                operator: Operator::Delete,
                target: Target::Motion(Motion::Search {
                    pattern: "bar".to_owned(),
                    backward: true,
                }),
            }
        );
        assert_eq!(command("/fop<BS>o<CR>").0, command("/foo<CR>").0);

        let keymap = Keymap::vim();
        let mut pending = Pending::new();
        for key in keys("/fop<BS>").expect("valid test notation") {
            assert_eq!(
                pending.feed(key, Mode::Normal, &keymap),
                Resolution::Pending
            );
        }
        assert_eq!(pending.keys(), keys("/fo").expect("valid test notation"));

        assert!(matches!(
            resolve_in(Mode::Normal, "/<BS>"),
            Resolution::Cancelled { .. }
        ));
        assert!(matches!(
            resolve_in(Mode::Normal, "/<CR>"),
            Resolution::Rejected { .. }
        ));
    }

    #[test]
    fn marks_are_commands_or_motions_as_their_grammar_requires() {
        assert_eq!(command("ma").0, Command::SetMark('a'));
        assert_eq!(
            command("`a").0,
            Command::Move(Motion::Mark {
                name: 'a',
                exact: true,
            })
        );
        // `'` and `` ` `` name the position before the latest jump, and go
        // through the motion path like any other mark — so they carry the same
        // linewise/characterwise distinction and take an operator.
        for (spec, name, exact) in [("''", '\'', false), ("``", '`', true)] {
            assert_eq!(
                command(spec).0,
                Command::Move(Motion::Mark { name, exact }),
                "{spec}"
            );
        }
        assert_eq!(
            command("d''").0,
            Command::Operate {
                operator: Operator::Delete,
                target: Target::Motion(Motion::Mark {
                    name: '\'',
                    exact: false,
                }),
            }
        );
        for (spec, name) in [
            ("`<lt>", '<'),
            ("`>", '>'),
            ("`[", '['),
            ("`]", ']'),
            ("`^", '^'),
        ] {
            assert_eq!(
                command(spec).0,
                Command::Move(Motion::Mark { name, exact: true }),
                "{spec}"
            );
        }
        assert!(matches!(
            resolve_in(Mode::Normal, "m<lt>"),
            Resolution::Rejected { .. }
        ));
    }

    #[test]
    fn insert_mode_falls_through_to_text() {
        for (mode, spec, expected) in [
            (Mode::Insert, "a", Command::InsertText('a')),
            (Mode::Insert, "3", Command::InsertText('3')),
            (Mode::Replace, "z", Command::InsertText('z')),
            (Mode::Insert, "<Esc>", Command::EnterNormal),
            (Mode::Insert, "<CR>", Command::InsertNewline),
        ] {
            match resolve_in(mode, spec) {
                Resolution::Command { command, count, .. } => {
                    assert_eq!(command, expected, "{spec}");
                    assert_eq!(count, None, "{spec}");
                }
                other => panic!("{spec:?} should resolve, got {other:?}"),
            }
        }
        assert!(matches!(
            resolve_in(Mode::Insert, "<C-q>"),
            Resolution::Rejected { .. }
        ));
        assert_eq!(command("i").0, Command::EnterInsert(InsertAt::Cursor));
        assert!(matches!(
            resolve_in(Mode::Visual(VisualKind::Char), "d"),
            Resolution::Command {
                command: Command::Operate {
                    target: Target::Selection,
                    ..
                },
                ..
            }
        ));
    }

    #[test]
    fn custom_keymaps_flow_through_the_grammar() {
        let mut keymap = Keymap::vim();
        keymap.bind_spec(Layer::Normal, "<C-a>", Binding::Motion(Motion::LastColumn));
        let mut pending = Pending::new();
        assert_eq!(
            pending.feed(Key::char('d'), Mode::Normal, &keymap),
            Resolution::Pending
        );
        assert!(matches!(
            pending.feed(Key::ctrl('a'), Mode::Normal, &keymap),
            Resolution::Command {
                command: Command::Operate {
                    operator: Operator::Delete,
                    target: Target::Motion(Motion::LastColumn),
                },
                ..
            }
        ));
    }
}

File: js/src/pending.js
// The pending-input parser: vi's command grammar.
//
// A trie alone cannot express vi's normal-mode syntax, because counts and
// operators *compose* rather than sequence:
//
//     [count] operator [count] motion | textobject
//     [count] command
//
// So this type holds the compositional state — counts, a pending operator, a
// pending object scope, a pending character argument — and defers the rest to
// the keymap. It touches no buffer, which makes the whole grammar testable as
// keys in, command out.

import * as C from './command.js';
import * as M from './motion.js';
import { L_INSERT, L_OPERATOR, PREFIX, UNBOUND, layerOf } from './keymap.js';
import { keyDigit, keyText } from './keys.js';

/** Resolutions. */
export const PENDING = 0;
export const COMMAND = 1;
export const REJECTED = 2;
export const CANCELLED = 3;

/** @typedef {import('./keys.js').Key} Key */
/** @typedef {{ r: number, command?: any, count?: number | null, keys?: Key[] }} Resolution */

const PENDING_RESULT = { r: PENDING };

export class Pending {
  constructor() {
    /** @type {Key[]} */
    this.keys = [];
    /** Start of the partial path through the current layer's bindings. */
    this.pathStart = 0;
    /** @type {number | null} */
    this.countBefore = null;
    /** @type {number | null} */
    this.countAfter = null;
    /** @type {number | null} */
    this.operator = null;
    /** @type {boolean | null} `i` or `a`, awaiting an object key. */
    this.scope = null;
    /** @type {any} */
    this.awaiting = null;
    /** @type {{ pattern: string, backward: boolean } | null} */
    this.search = null;
  }

  /** Nothing accumulated — the next key starts a fresh command. @returns {boolean} */
  get idle() {
    return (
      this.pathStart === this.keys.length &&
      this.countBefore === null &&
      this.countAfter === null &&
      this.operator === null &&
      this.scope === null &&
      this.awaiting === null &&
      this.search === null
    );
  }

  reset() {
    this.keys = [];
    this.pathStart = 0;
    this.countBefore = null;
    this.countAfter = null;
    this.operator = null;
    this.scope = null;
    this.awaiting = null;
    this.search = null;
  }

  /**
   * Feed one key. `mode` selects the grammar: command modes get counts and
   * operators, insert and replace get a straight keymap lookup with unbound
   * printable keys falling through to text.
   * @param {Key} key
   * @param {number} mode
   * @param {import('./keymap.js').Keymap} keymap
   * @returns {Resolution}
   */
  feed(key, mode, keymap) {
    if (!C.isCommandMode(mode)) {
      this.keys.push(key);
      return this.#insert(key, keymap);
    }

    // `<Esc>` abandons a partial sequence. When nothing is pending it falls
    // through to the keymap, where it means "return to normal mode".
    const idle = this.idle;
    const fresh = this.pathStart === this.keys.length;
    this.keys.push(key);
    if (key === '<Esc>' && !idle) return this.#end(CANCELLED);

    if (this.search !== null) return this.#feedSearch(key);

    if (this.awaiting !== null) {
      const text = keyText(key);
      return text === null ? this.#end(REJECTED) : this.#await(this.awaiting, text);
    }

    if (this.scope !== null) {
      const object = keymap.object(key);
      return object === undefined ? this.#end(REJECTED) : this.#object(this.scope, object);
    }

    // Counts accumulate only at the start of a key path, and a leading `0` is
    // the first-column motion rather than a digit — the classic ambiguity.
    if (fresh) {
      const digit = keyDigit(key);
      if (digit >= 0) {
        const before = this.operator === null;
        const slot = before ? this.countBefore : this.countAfter;
        if (!(digit === 0 && slot === null)) {
          const value = (slot ?? 0) * 10 + digit;
          if (before) this.countBefore = value;
          else this.countAfter = value;
          this.pathStart = this.keys.length;
          return PENDING_RESULT;
        }
      }
    }

    const layer = this.operator !== null ? L_OPERATOR : layerOf(mode);
    const found = keymap.walk(layer, this.keys, this.pathStart);
    if (found === PREFIX) return PENDING_RESULT;
    if (found === UNBOUND) return this.#end(REJECTED);
    return this.#apply(/** @type {import('./keymap.js').Binding} */ (found), mode);
  }

  /**
   * @param {Key} key
   * @param {import('./keymap.js').Keymap} keymap
   * @returns {Resolution}
   */
  #insert(key, keymap) {
    const walked = keymap.walk(L_INSERT, this.keys, this.pathStart);
    if (walked === PREFIX) return PENDING_RESULT;
    if (walked !== UNBOUND) {
      // Operators and object scopes are meaningless while inserting.
      const found = /** @type {import('./keymap.js').Binding} */ (walked);
      return found.b === C.B_COMMAND ? this.#finish(found.command) : this.#end(REJECTED);
    }
    // An unbound printable key is text.
    const text = this.keys.length - this.pathStart === 1 ? keyText(key) : null;
    return text === null ? this.#end(REJECTED) : this.#finish({ c: C.INSERT_TEXT, text });
  }

  /**
   * @param {import('./keymap.js').Binding} binding
   * @param {number} mode
   * @returns {Resolution}
   */
  #apply(binding, mode) {
    this.pathStart = this.keys.length;
    switch (binding.b) {
      case C.B_COMMAND:
        // An operator needs a target; a plain command is not one.
        return this.operator === null ? this.#finish(binding.command) : this.#end(REJECTED);
      case C.B_MOTION:
        return this.#finish(this.#withMotion(binding.motion));
      case C.B_OPERATOR: {
        // In visual mode an operator applies to the selection at once.
        if (C.isVisual(mode)) {
          return this.#finish({
            c: C.OPERATE,
            operator: binding.operator,
            target: SELECTION,
          });
        }
        if (this.operator === null) {
          this.operator = binding.operator;
          return PENDING_RESULT;
        }
        // A doubled operator is linewise: `dd`, `cc`, `yy`.
        if (this.operator === binding.operator) {
          return this.#finish({ c: C.OPERATE, operator: binding.operator, target: CURRENT_ROW });
        }
        return this.#end(REJECTED);
      }
      case C.B_SCOPE:
        if (this.operator === null && !C.isVisual(mode)) return this.#end(REJECTED);
        this.scope = binding.around;
        return PENDING_RESULT;
      case C.B_AWAIT:
        if (
          binding.await === C.AWAIT_SURROUND_TARGET &&
          this.operator !== C.CHANGE &&
          this.operator !== C.DELETE
        ) {
          // `Yank` is where `ys` can hook in when its target grammar is built.
          return this.#end(REJECTED);
        }
        this.awaiting = binding;
        return PENDING_RESULT;
      default:
        this.search = { pattern: '', backward: binding.backward };
        return PENDING_RESULT;
    }
  }

  /**
   * @param {Key} key
   * @returns {Resolution}
   */
  #feedSearch(key) {
    const search = /** @type {{ pattern: string, backward: boolean }} */ (this.search);
    if (key === '<CR>') {
      if (search.pattern === '') return this.#end(REJECTED);
      this.search = null;
      return this.#finish(
        this.#withMotion({ k: M.SEARCH, pattern: search.pattern, backward: search.backward }),
      );
    }
    if (key === '<BS>') {
      if (search.pattern === '') return this.#end(CANCELLED);
      search.pattern = search.pattern.slice(0, -1);
      // `keys` doubles as showcmd. Keep its displayed pattern in step with
      // editing; replaying this normalised spelling has the same result as
      // replaying the literal backspace.
      this.keys.length -= 2;
      return PENDING_RESULT;
    }
    const text = keyText(key);
    if (text === null) return this.#end(REJECTED);
    search.pattern += text;
    return PENDING_RESULT;
  }

  /**
   * @param {any} awaiting
   * @param {string} ch
   * @returns {Resolution}
   */
  #await(awaiting, ch) {
    this.awaiting = null;
    const free = this.operator === null;
    switch (awaiting.await) {
      case C.AWAIT_FIND:
        return this.#finish(
          this.#withMotion({ k: M.FIND, target: ch, backward: awaiting.backward, till: awaiting.till }),
        );
      // These take no target, so a pending operator is a syntax error.
      case C.AWAIT_REPLACE_CHAR:
        return free ? this.#finish({ c: C.REPLACE_CHAR, text: ch }) : this.#end(REJECTED);
      case C.AWAIT_RECORD:
        return free ? this.#finish({ c: C.RECORD_MACRO, register: ch }) : this.#end(REJECTED);
      case C.AWAIT_PLAY:
        return free ? this.#finish({ c: C.PLAY_MACRO, register: ch }) : this.#end(REJECTED);
      case C.AWAIT_SET_MARK:
        return free && ch >= 'a' && ch <= 'z'
          ? this.#finish({ c: C.SET_MARK, name: ch })
          : this.#end(REJECTED);
      // `'` and `` ` `` name the position before the latest jump. They go
      // through the motion path like any other mark, so `d''` is linewise and
      // ``d`` `` characterwise, as `d'a` and ``d`a`` are.
      case C.AWAIT_GOTO_MARK:
        return (ch >= 'a' && ch <= 'z') || MARK_NAMES.includes(ch)
          ? this.#finish(this.#withMotion({ k: M.MARK, name: ch, exact: awaiting.exact }))
          : this.#end(REJECTED);
      case C.AWAIT_SURROUND_TARGET:
        if (this.operator === C.DELETE) return this.#finish({ c: C.DELETE_SURROUND, target: ch });
        if (this.operator === C.CHANGE) {
          this.awaiting = { await: C.AWAIT_SURROUND_TO, from: ch };
          return PENDING_RESULT;
        }
        return this.#end(REJECTED);
      case C.AWAIT_SURROUND_TO:
        return this.operator === C.CHANGE
          ? this.#finish({ c: C.CHANGE_SURROUND, from: awaiting.from, to: ch })
          : this.#end(REJECTED);
      default:
        return free ? this.#finish({ c: C.SURROUND_SELECTION, delimiter: ch }) : this.#end(REJECTED);
    }
  }

  /**
   * @param {boolean} around
   * @param {import('./motion.js').TextObject} object
   * @returns {Resolution}
   */
  #object(around, object) {
    this.scope = null;
    return this.#finish(
      this.operator === null
        ? { c: C.SELECT_OBJECT, around, object }
        : { c: C.OPERATE, operator: this.operator, target: { t: C.T_OBJECT, around, object } },
    );
  }

  /**
   * A motion is a movement on its own, or an operator's target.
   * @param {import('./motion.js').Motion} motion
   * @returns {object}
   */
  #withMotion(motion) {
    return this.operator === null
      ? { c: C.MOVE, motion }
      : { c: C.OPERATE, operator: this.operator, target: { t: C.T_MOTION, motion } };
  }

  /**
   * vi multiplies the two counts: `2d3w` deletes six words.
   * @returns {number | null}
   */
  #count() {
    if (this.countBefore === null && this.countAfter === null) return null;
    return (this.countBefore ?? 1) * (this.countAfter ?? 1);
  }

  /**
   * @param {object} command
   * @returns {Resolution}
   */
  #finish(command) {
    const count = this.#count();
    const keys = this.keys;
    this.reset();
    return { r: COMMAND, command, count, keys };
  }

  /**
   * @param {number} r
   * @returns {Resolution}
   */
  #end(r) {
    const keys = this.keys;
    this.reset();
    return { r, keys };
  }
}

const SELECTION = { t: C.T_SELECTION };
const CURRENT_ROW = { t: C.T_CURRENT_ROW };
const MARK_NAMES = ['<', '>', '[', ']', '^', "'", '`'];
